### PR TITLE
fix(sdk): during file upload, handle Azure HTTP errors better

### PR DESCRIPTION
**Pull request by speezepearson**
Description
-----------

When `InternalApi.upload_file()` tries to upload to Azure, and something goes wrong, it mis-handles the AzureError, masking it with an AttributeError, which is then not retried.

I demonstrated this by running a local server which 500ed every request, and forcing `upload_file` to (a) always go down the Azure code-path, and (b) hit my local server instead of the real Azure.

<details><summary>Code patch:</summary>

```diff
--- a/wandb/sdk/internal/internal_api.py
+++ b/wandb/sdk/internal/internal_api.py
@@ -1952,7 +1952,9 @@ class Api:
         response: Optional[requests.Response] = None
         progress = Progress(file, callback=callback)
         try:
-            if "x-ms-blob-type" in extra_headers and self._azure_blob_module:
+            if True:
+                import urllib.parse
+                url = urllib.parse.urlparse(url)._replace(scheme="http", netloc="localhost:8333").geturl()
                 self.upload_file_azure(url, progress, extra_headers)
```

</details>

What it _currently_ does is try to access an attribute on the exception that doesn't exist:
<details><summary>Before:</summary>

```
~/s/client (main) $ python3 -c 'import wandb; wandb.init()'
<...snip...>
wandb: Waiting for W&B process to finish... (success).
wandb: ERROR Error uploading "wandb-metadata.json": CommError, 'Response' object has no attribute 'read'
wandb: ERROR Error uploading "requirements.txt": CommError, 'Response' object has no attribute 'read'
wandb: ERROR Error uploading "wandb-summary.json": CommError, 'Response' object has no attribute 'read'
wandb: ERROR Error uploading "config.yaml": CommError, 'Response' object has no attribute 'read'
wandb: Synced rosy-sun-6: https://app.qa.wandb.ai/spencerpearson-wandb/client/runs/3qxuh134
wandb: Synced 4 W&B file(s), 0 media file(s), 0 artifact file(s) and 0 other file(s)
wandb: Find logs at: ./wandb/run-20221129_142838-3qxuh134/logs
```

`debug-internal.log`:
```
2022-11-29 14:28:39,631 ERROR   Thread-16 :66791 [internal_api.py:upload_file():1927] upload_file exception http://localhost:8333/wandb-qa.appspot.com/spencerpearson-wandb/client/3qxuh134/wandb-metadata.json?Expires=1669847319&GoogleAccessId=wandb-qa%40appspot.gserviceaccount.com&Signature=ZSSbA%2BGsdR8%2BIs9RXyO5CpE5ue%2BGlo7ZTM668AC3zHRoAOCP29DvQPGgVX5IH0nwb72S5jzxn1wf%2B7OuS0RUeeS1gCrMLA%2B2OzParq4dNSzhy3V53J3S25d0qOmRNSmQjFEQXSyvezF4Ke%2BnhXpz4TfZoppchL2539JeleXoNipjOcgqpg3oZ%2FJPYH9DO%2Fao5pgVyAO8p2gD8bIDl5NK%2FF6QQw7CJf9tR7wIyvMYD4Ful7WfKpq6lhKekE6KtYaWJCIYu2xfbjPh9cxy8oe%2FJ38KpotEcJGG%2BLRAxsTfCo6Fkye%2B5twbqGURzAgaqgA4eIn4%2FzMNUuB6J8HFVXrtGA%3D%3D: Internal Server Error
ErrorCode:None
2022-11-29 14:28:39,631 ERROR   Thread-16 :66791 [internal_api.py:upload_file():1929] upload_file request headers:
2022-11-29 14:28:39,631 ERROR   Thread-16 :66791 [upload_job.py:push():146] Failed to upload file: /var/folders/wp/vvlmjhbd6q3gdr_l4grs6cyc0000gn/T/tmpq7d8gabjwandb/2a0zla6w-wandb-metadata.json
Traceback (most recent call last):
  File "/Users/pears/src/client/wandb/sdk/internal/internal_api.py", line 1874, in upload_file_azure
    client.upload_blob(
  File "/Users/pears/.pyenv/versions/example/lib/python3.9/site-packages/azure/core/tracing/decorator.py", line 78, in wrapper_use_tracer
    return func(*args, **kwargs)
  File "/Users/pears/.pyenv/versions/example/lib/python3.9/site-packages/azure/storage/blob/_blob_client.py", line 731, in upload_blob
    return upload_block_blob(**options)
  File "/Users/pears/.pyenv/versions/example/lib/python3.9/site-packages/azure/storage/blob/_upload_helpers.py", line 197, in upload_block_blob
    process_storage_error(error)
  File "/Users/pears/.pyenv/versions/example/lib/python3.9/site-packages/azure/storage/blob/_shared/response_handlers.py", line 181, in process_storage_error
    exec("raise error from None")   # pylint: disable=exec-used # nosec
  File "<string>", line 1, in <module>
azure.core.exceptions.HttpResponseError: Internal Server Error
ErrorCode:None

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/pears/src/client/wandb/sdk/internal/internal_api.py", line 1917, in upload_file
    self.upload_file_azure(url, progress, extra_headers)
  File "/Users/pears/src/client/wandb/sdk/internal/internal_api.py", line 1887, in upload_file_azure
    raise requests.exceptions.RequestException(e.message, response=response)
requests.exceptions.RequestException: Internal Server Error
ErrorCode:None

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/pears/src/client/wandb/apis/normalize.py", line 26, in wrapper
    return func(*args, **kwargs)
  File "/Users/pears/src/client/wandb/sdk/lib/retry.py", line 190, in wrapped_fn
    return retrier(*args, **kargs)
  File "/Users/pears/src/client/wandb/sdk/lib/retry.py", line 113, in __call__
    result = self._call_fn(*args, **kwargs)
  File "/Users/pears/src/client/wandb/sdk/internal/internal_api.py", line 1930, in upload_file
    response_content = e.response.content if e.response is not None else ""
  File "/Users/pears/.pyenv/versions/example/lib/python3.9/site-packages/requests/models.py", line 899, in content
    self._content = b"".join(self.iter_content(CONTENT_CHUNK_SIZE)) or b""
  File "/Users/pears/.pyenv/versions/example/lib/python3.9/site-packages/requests/models.py", line 828, in generate
    chunk = self.raw.read(chunk_size)
AttributeError: 'Response' object has no attribute 'read'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/pears/src/client/wandb/filesync/upload_job.py", line 137, in push
    self._api.upload_file_retry(
  File "/Users/pears/src/client/wandb/apis/normalize.py", line 62, in wrapper
    raise CommError(message, err).with_traceback(sys.exc_info()[2])
  File "/Users/pears/src/client/wandb/apis/normalize.py", line 26, in wrapper
    return func(*args, **kwargs)
  File "/Users/pears/src/client/wandb/sdk/lib/retry.py", line 190, in wrapped_fn
    return retrier(*args, **kargs)
  File "/Users/pears/src/client/wandb/sdk/lib/retry.py", line 113, in __call__
    result = self._call_fn(*args, **kwargs)
  File "/Users/pears/src/client/wandb/sdk/internal/internal_api.py", line 1930, in upload_file
    response_content = e.response.content if e.response is not None else ""
  File "/Users/pears/.pyenv/versions/example/lib/python3.9/site-packages/requests/models.py", line 899, in content
    self._content = b"".join(self.iter_content(CONTENT_CHUNK_SIZE)) or b""
  File "/Users/pears/.pyenv/versions/example/lib/python3.9/site-packages/requests/models.py", line 828, in generate
    chunk = self.raw.read(chunk_size)
wandb.errors.CommError: 'Response' object has no attribute 'read'
```

</details>

What it _should_ do is wrap the exception in a TransientError, which our retry logic will recognize:
<details><summary>After:</summary>

```
~/s/client (spencerpearson/azure-response-fix) $ python3 -c 'import wandb; wandb.init()'
<...snip...>
wandb: Waiting for W&B process to finish... (success).
wandb: Network error (TransientError), entering retry loop.
^C  (<-- it was hanging in the retry loop, as it should)
```

`debug-internal.log`:

```
2022-11-29 14:29:06,756 ERROR   Thread-17 :67613 [internal_api.py:upload_file():1968] upload_file exception http://localhost:8333/wandb-qa.appspot.com/spencerpearson-wandb/client/3dvsxpdc/requirements.txt?Expires=1669847345&GoogleAccessId=wandb-qa%40appspot.gserviceaccount.com&Signature=j4k9O9nLqSVaXVqXNfeIJWvLbHTbzMkKZO5NWG83gLpT%2B0viThchFlylHg%2Flv8iIL%2FvvnscOnCrOVNoHhbhcrkCxn1dull8J1MZVMpUOP8ZoeH6YUIroXyeXiVqjkCT4TOnuEaIPavVCplVzXlMW%2FNXbdwyOEy63HrncKmE3CBHCReJgGf4%2BKlFqKN55J2%2BdBhJVBPqCmhbZkIF6ZtIBadgoLPcPXJ%2FrZdpeTOFZhypfhDUd0x0mrEuby2YvaAXJmXTwmom%2BAmyuGPB0HmxuL7MR59v5F5z0pBs%2BVsXnugHoLlvw9wCDNYdzhc07wZKavkc8TECHR%2BFUuZnPMso4ig%3D%3D: Internal Server Error
ErrorCode:None
2022-11-29 14:29:06,756 ERROR   Thread-17 :67613 [internal_api.py:upload_file():1970] upload_file request headers:
2022-11-29 14:29:06,756 ERROR   Thread-17 :67613 [internal_api.py:upload_file():1972] upload_file response body: b''
2022-11-29 14:29:06,756 ERROR   Thread-17 :67613 [retry.py:__call__():154] Retry attempt failed:
Traceback (most recent call last):
  File "/Users/pears/src/client/wandb/sdk/internal/internal_api.py", line 1915, in upload_file_azure
    client.upload_blob(
  File "/Users/pears/.pyenv/versions/example/lib/python3.9/site-packages/azure/core/tracing/decorator.py", line 78, in wrapper_use_tracer
    return func(*args, **kwargs)
  File "/Users/pears/.pyenv/versions/example/lib/python3.9/site-packages/azure/storage/blob/_blob_client.py", line 731, in upload_blob
    return upload_block_blob(**options)
  File "/Users/pears/.pyenv/versions/example/lib/python3.9/site-packages/azure/storage/blob/_upload_helpers.py", line 197, in upload_block_blob
    process_storage_error(error)
  File "/Users/pears/.pyenv/versions/example/lib/python3.9/site-packages/azure/storage/blob/_shared/response_handlers.py", line 181, in process_storage_error
    exec("raise error from None")   # pylint: disable=exec-used # nosec
  File "<string>", line 1, in <module>
azure.core.exceptions.HttpResponseError: Internal Server Error
ErrorCode:None

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/pears/src/client/wandb/sdk/internal/internal_api.py", line 1958, in upload_file
    self.upload_file_azure(url, progress, extra_headers)
  File "/Users/pears/src/client/wandb/sdk/internal/internal_api.py", line 1928, in upload_file_azure
    raise requests.exceptions.RequestException(e.message, response=response)
requests.exceptions.RequestException: Internal Server Error
ErrorCode:None

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/pears/src/client/wandb/sdk/lib/retry.py", line 113, in __call__
    result = self._call_fn(*args, **kwargs)
  File "/Users/pears/src/client/wandb/sdk/internal/internal_api.py", line 1992, in upload_file
    raise _e.with_traceback(sys.exc_info()[2])
  File "/Users/pears/src/client/wandb/sdk/internal/internal_api.py", line 1958, in upload_file
    self.upload_file_azure(url, progress, extra_headers)
  File "/Users/pears/src/client/wandb/sdk/internal/internal_api.py", line 1928, in upload_file_azure
    raise requests.exceptions.RequestException(e.message, response=response)
wandb.sdk.lib.retry.TransientError: None
2022-11-29 14:29:06,782 ERROR   Thread-19 :67613 [internal_api.py:upload_file():1968] upload_file exception http://localhost:8333/wandb-qa.appspot.com/spencerpearson-wandb/client/3dvsxpdc/wandb-summary.json?Expires=1669847345&GoogleAccessId=wandb-qa%40appspot.gserviceaccount.com&Signature=rfZscEf6YY8tdBsRiqr%2BacDRS%2FndQAMWnSKoiHIkbFpXxHm4dSC5b%2Fsi4lPc7C5BpYH7YyuzBSnuI%2FbC0bnNzcBIaKNkNijgrwELr9tbDDPKs%2FLep5W0xxHN5Y5Vt%2BfsZCXb1QUqX%2BgFOI8TfXCJ1quagV7b1XJpig15VWOC1sBINsopYPrFCHmpSahJK6MzJSbX%2BevrwWpnNVK7nByrMfR5cOP6Gsazg7ANUCQyAlL9kYOOTHydyaZqj9%2BsmMCL%2FwB4jGGYdmvPIHkCEpReZtbXtQQy6%2BdecnPFtuqYsTqdno3mNV%2FMP0Pm8qzBrym32DBG9bI%2F7D0OOv35WWUxXw%3D%3D: Internal Server Error
ErrorCode:None
2022-11-29 14:29:06,782 ERROR   Thread-19 :67613 [internal_api.py:upload_file():1970] upload_file request headers:
2022-11-29 14:29:06,782 ERROR   Thread-19 :67613 [internal_api.py:upload_file():1972] upload_file response body: b''
2022-11-29 14:29:06,885 ERROR   Thread-18 :67613 [internal_api.py:upload_file():1968] upload_file exception http://localhost:8333/wandb-qa.appspot.com/spencerpearson-wandb/client/3dvsxpdc/config.yaml?Expires=1669847345&GoogleAccessId=wandb-qa%40appspot.gserviceaccount.com&Signature=hsR%2FQzH5OiRQOADhCZV1Z3cVziISY9txF3NhMuu2A3%2FpW8v6I6SuudXnYk9gKObvAXPltze3aGkPrpvDCR9C%2FgEPpIjuhOjLJ6ePvk0psRsEaTmmAj1CbswAv8zavUvdb1fFGD5fpyk%2FB37%2FZ3ZpdAAxfh4XjFHUqqk084RDKuir3jpcfTP49iymvw11kVPpSIVYTd%2Bow7hmXd8X4zkLNeGki3VIpcTsow6GvOnaPdG6h21vqEchiilqNy%2Bg17glcyqm14aeVaQoHBzhm1ETsLo4ZgooEaCeLnedv2i66u4RMC%2Fj6C%2Fn5diHWwY2n8OPsF0iShrkF0s5Uykuv7n0%2BA%3D%3D: Internal Server Error
ErrorCode:None
2022-11-29 14:29:06,885 ERROR   Thread-18 :67613 [internal_api.py:upload_file():1970] upload_file request headers:
2022-11-29 14:29:06,885 ERROR   Thread-18 :67613 [internal_api.py:upload_file():1972] upload_file response body: b''
2022-11-29 14:29:07,722 DEBUG   HandlerThread:67613 [handler.py:handle_request():139] handle_request: poll_exit
2022-11-29 14:29:07,722 DEBUG   SenderThread:67613 [sender.py:send_request():317] send_request: poll_exit
2022-11-29 14:29:08,242 ERROR   Thread-16 :67613 [internal_api.py:upload_file():1968] upload_file exception http://localhost:8333/wandb-qa.appspot.com/spencerpearson-wandb/client/3dvsxpdc/wandb-metadata.json?Expires=1669847344&GoogleAccessId=wandb-qa%40appspot.gserviceaccount.com&Signature=SElDRod5qOtx07KCuqWO2JDwDBjp%2FG4%2BOgn6u2bsptiDjtx%2BJW%2Bsd1bomR5zvQCfdiQnhJ9%2F9OJHmyZvIgzO93Zq1NHyWOUTsCpTLy9MkVz8rmXikECJHMoRyj74pmQD2isNyyi4xuOfmX5oCA6HazVQe3imPEyon1xdhb4Cq529LDMGT8efkMuphPPJGVRUu%2BbqARxzD1EN1KEkSWxZsLMXE7ECPjZ9x5dLSu8aMQ4mPTQeMWW%2Bg9QBDGAc%2FvF2Eky2%2F9oEeIbQhV5mbhI56bEp6YV4OYutsGQQdSb58pHB6habGU1GFB%2F21qc40F5FmJZhbJGo6QmXz5%2BecMtHKA%3D%3D: Internal Server Error
ErrorCode:None
2022-11-29 14:29:08,242 ERROR   Thread-16 :67613 [internal_api.py:upload_file():1970] upload_file request headers:
2022-11-29 14:29:08,242 ERROR   Thread-16 :67613 [internal_api.py:upload_file():1972] upload_file response body: b''
2022-11-29 14:29:08,724 DEBUG   HandlerThread:67613 [handler.py:handle_request():139] handle_request: poll_exit
2022-11-29 14:29:08,724 DEBUG   SenderThread:67613 [sender.py:send_request():317] send_request: poll_exit
2022-11-29 14:29:08,890 ERROR   Thread-19 :67613 [internal_api.py:upload_file():1968] upload_file exception http://localhost:8333/wandb-qa.appspot.com/spencerpearson-wandb/client/3dvsxpdc/wandb-summary.json?Expires=1669847345&GoogleAccessId=wandb-qa%40appspot.gserviceaccount.com&Signature=rfZscEf6YY8tdBsRiqr%2BacDRS%2FndQAMWnSKoiHIkbFpXxHm4dSC5b%2Fsi4lPc7C5BpYH7YyuzBSnuI%2FbC0bnNzcBIaKNkNijgrwELr9tbDDPKs%2FLep5W0xxHN5Y5Vt%2BfsZCXb1QUqX%2BgFOI8TfXCJ1quagV7b1XJpig15VWOC1sBINsopYPrFCHmpSahJK6MzJSbX%2BevrwWpnNVK7nByrMfR5cOP6Gsazg7ANUCQyAlL9kYOOTHydyaZqj9%2BsmMCL%2FwB4jGGYdmvPIHkCEpReZtbXtQQy6%2BdecnPFtuqYsTqdno3mNV%2FMP0Pm8qzBrym32DBG9bI%2F7D0OOv35WWUxXw%3D%3D: Internal Server Error
ErrorCode:None
2022-11-29 14:29:08,890 ERROR   Thread-19 :67613 [internal_api.py:upload_file():1970] upload_file request headers:
2022-11-29 14:29:08,890 ERROR   Thread-19 :67613 [internal_api.py:upload_file():1972] upload_file response body: b''
2022-11-29 14:29:09,143 ERROR   Thread-17 :67613 [internal_api.py:upload_file():1968] upload_file exception http://localhost:8333/wandb-qa.appspot.com/spencerpearson-wandb/client/3dvsxpdc/requirements.txt?Expires=1669847345&GoogleAccessId=wandb-qa%40appspot.gserviceaccount.com&Signature=j4k9O9nLqSVaXVqXNfeIJWvLbHTbzMkKZO5NWG83gLpT%2B0viThchFlylHg%2Flv8iIL%2FvvnscOnCrOVNoHhbhcrkCxn1dull8J1MZVMpUOP8ZoeH6YUIroXyeXiVqjkCT4TOnuEaIPavVCplVzXlMW%2FNXbdwyOEy63HrncKmE3CBHCReJgGf4%2BKlFqKN55J2%2BdBhJVBPqCmhbZkIF6ZtIBadgoLPcPXJ%2FrZdpeTOFZhypfhDUd0x0mrEuby2YvaAXJmXTwmom%2BAmyuGPB0HmxuL7MR59v5F5z0pBs%2BVsXnugHoLlvw9wCDNYdzhc07wZKavkc8TECHR%2BFUuZnPMso4ig%3D%3D: Internal Server Error
<...snip...>
```

</details>

### CODE REVIEW

File: wandb/sdk/internal/internal_api.py [hunk 1]
```diff
@@ -96,6 +97,45 @@ class DefaultSettings(TypedDict):
 #     def __getitem__(self, name: str) -> Any: ...
 
 
+def _guess_response_content(response: Any) -> bytes:
```
Comment (kptkin) [inline wandb/sdk/internal/internal_api.py:100]: wow... that sucks that we have to do this :(
but based on your description there is nothing better here

File: wandb/sdk/internal/internal_api.py [hunk 2]
```diff
@@ -96,6 +97,45 @@ class DefaultSettings(TypedDict):
...
+            else _guess_response_content(response.internal_response)
+            if hasattr(response, "internal_response")
+            else response.content
+            if hasattr(response, "content")
+            else response.text
+            if hasattr(response, "text")
+            else response.read()
+            if hasattr(response, "read")
+            else "<unknown>"
+        )
```
[the remark this answers is not in the document]
Comment (speezepearson) [reply wandb/sdk/internal/internal_api.py]: Yeah, good call. I think was was more readable _before_ linting. 😛 
Fixed.

File: wandb/sdk/internal/internal_api.py (hunk 1 above)
Comment (speezepearson) [reply wandb/sdk/internal/internal_api.py:100]: Yeah :( Bummer.

**Comment by speezepearson** [comment]
Never mind! #4539 just yanks out the problematic logic altogether: our handling of the `internal_response` is busted, and the info we're trying to get out of it (the response content) shows up in `e.message` anyway.

**Pull request closed without merging**